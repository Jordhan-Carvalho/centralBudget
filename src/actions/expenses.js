import db from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

    
  export const startAddExpense = (expenseData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0,
      } = expenseData;
      const expense = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/expenses`).push(expense);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

  export const startRemoveExpense = (expense) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = expense.id;
  try{
  await db.ref(`users/${uid}/expenses/${id}`).remove();
        dispatch(removeExpense(expense));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  });

  export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/expenses/${id}`).update(updates);
        dispatch(editExpense(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  // export const startSetExpenses
  export const startSetExpenses = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all expense data once
  //2. Parse the data to array
  const expenses = []; 
  await db.ref(`users/${uid}/expenses`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      };
      };