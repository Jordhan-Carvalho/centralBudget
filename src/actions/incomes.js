import db from '../firebase/firebase';

// ADD_INCOME
export const addIncome = (income) => ({
    type: 'ADD_INCOME',
    income
  });

    
  export const startAddIncome = (incomeData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
      } = incomeData;
      const income = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/incomes`).push(income);
      dispatch(addIncome({
        id: ref.key,
        ...income
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeIncome = ({ id } = {}) => ({
    type: 'REMOVE_INCOME',
    id
  });

  export const startRemoveIncome = (income) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = income.id;
  try{
  await db.ref(`users/${uid}/incomes/${id}`).remove();
        dispatch(removeIncome(income));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editIncome = (id, updates) => ({
    type: 'EDIT_INCOME',
    id,
    updates
  });

  export const startEditIncome = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/incomes/${id}`).update(updates);
        dispatch(editIncome(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setIncomes = (incomes) => ({
    type: 'SET_INCOMES',
    incomes
  });

  // export const startSetExpenses
  export const startSetIncomes = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all income data once
  //2. Parse the data to array
  const incomes = []; 
  await db.ref(`users/${uid}/incomes`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                incomes.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setIncomes(incomes));
      };
      };