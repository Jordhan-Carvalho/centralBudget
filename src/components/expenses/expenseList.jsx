import React from "react";
import { connect } from "react-redux";
import ListItem from "../common/listItem";
import selectVisible from "../../selectors/visibleTrans";

export const ExpenseList = props => {
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <ListItem type={"expense"} expenses={props.expenses} />
      )}
    </div>
  );
};

// redux call
const mapStateToProps = state => {
  return {
    expenses: selectVisible(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
