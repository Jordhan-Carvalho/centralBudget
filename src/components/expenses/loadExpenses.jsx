import React, { Component } from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./expenseList";
import ListFilters from "../common/listFilters";
import TransSumary from "../common/transSumary";

class LoadExpenses extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Expense List</h1>
        <hr />
        <ListFilters />
        <TransSumary type={"expense"} />
        <hr />
        <ExpenseList />
        <hr />
        <Link to="/create">
          <button className="btn btn-dark">Add Expense</button>
        </Link>
      </React.Fragment>
    );
  }
}

export default LoadExpenses;
