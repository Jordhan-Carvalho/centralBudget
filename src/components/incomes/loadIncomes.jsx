import React, { Component } from "react";
import { Link } from "react-router-dom";
import IncomeList from "./incomeList";
import ListFilters from "../common/listFilters";
import TransSumary from "../common/transSumary";

class LoadIncome extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Incomes List</h1>
        <hr />
        <ListFilters />
        <TransSumary type={"income"} />
        <hr />
        <IncomeList />
        <hr />
        <Link to="/createinc">
          <button className="btn btn-dark">Add Income</button>
        </Link>
      </React.Fragment>
    );
  }
}

export default LoadIncome;
