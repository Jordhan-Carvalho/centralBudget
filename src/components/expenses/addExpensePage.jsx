import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddExpense } from "../../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/expenses");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Expense</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
