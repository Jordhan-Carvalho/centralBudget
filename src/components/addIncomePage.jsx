import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "./common/form";
import { startAddIncome } from "../actions/incomes";

export class AddIncomePage extends Component {
  onSubmit = income => {
    this.props.startAddIncome(income);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Income</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddIncome: income => dispatch(startAddIncome(income))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddIncomePage);
