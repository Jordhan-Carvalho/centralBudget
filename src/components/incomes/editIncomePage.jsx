import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startEditIncome, startRemoveIncome } from "../../actions/incomes";

export class EditIncomePage extends Component {
  onRemove = () => {
    this.props.startRemoveIncome(this.props.income);
    this.props.history.push("/incomes");
  };

  onSubmit = ({ description, note, amount, createdAt }) => {
    const income = { description, note, amount, createdAt };
    this.props.startEditIncome(this.props.income.id, income);
    this.props.history.push("/incomes");
  };

  render() {
    return (
      <div className="form-group">
        <h1>Edit {this.props.income.description}</h1>
        <div className="form-group">
          <Form expense={this.props.income} onSubmit={this.onSubmit} />
        </div>
        <div className="form-group">
          <button className="btn btn-danger" onClick={this.onRemove}>
            Remove
          </button>
        </div>
      </div>
    );
  }
}

// maps redux

const mapDispatchToProps = dispatch => {
  return {
    startEditIncome: (incomeID, income) =>
      dispatch(startEditIncome(incomeID, income)),
    startRemoveIncome: income => dispatch(startRemoveIncome(income))
  };
};

const mapStateToProps = (state, props) => {
  return {
    income: state.incomes.find(income => {
      return income.id === props.match.params.id;
    })
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditIncomePage);
