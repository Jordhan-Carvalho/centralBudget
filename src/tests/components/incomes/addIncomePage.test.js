import React from 'react';
import { shallow } from 'enzyme';
import { AddIncomePage } from '../../../components/incomes/addIncomePage';
import expenses from '../../fixtures/expenses';

// using lifecycle methods (global) refactoring
let startAddIncome, history, wrapper;
beforeEach(() => {
    startAddIncome = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddIncomePage startAddIncome={startAddIncome} history={history} />);
});
//

test('should render AddIncomePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

//test in case of submiting everything right
test('should handle onSubmit', () => {
    wrapper.find('Form').prop('onSubmit')(expenses[0]);
    //check if spies (mock fn) are ij
    expect(history.push).toHaveBeenLastCalledWith('/incomes');
    expect(startAddIncome).toHaveBeenLastCalledWith(expenses[0]);
});