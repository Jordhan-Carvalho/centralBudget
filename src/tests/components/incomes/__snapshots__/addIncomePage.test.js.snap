// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddIncomePage correctly 1`] = `
<Fragment>
  <h1>
    Add Income
  </h1>
  <Form
    isIncome={true}
    onSubmit={[Function]}
  />
</Fragment>
`;
